src/rust/engine/engine_pyo3/build.rs

```diff
@@ -0,0 +1,19 @@
+// Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+fn main() {
+  // NB: The native extension only works with the Python interpreter version it was built with
+  // (e.g. Python 3.7 vs 3.8).
+  println!("cargo:rerun-if-env-changed=PY");
+
+  if cfg!(target_os = "macos") {
```

Comment:
FYI this bit won't be necessary on the upcoming PyO3 0.14 (as it'll be done upstream for you).